materialize-boilerplate/apply.go

```diff
@@ -0,0 +1,232 @@
+package boilerplate
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	pf "github.com/estuary/flow/go/protocols/flow"
+	pm "github.com/estuary/flow/go/protocols/materialize"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	maxConcurrentUpdateActions = 5
+)
+
+// ActionApplyFn is a callback that will be executed to carry out some action to achieve a change to
+// a materialized resource.
+type ActionApplyFn func(context.Context) error
+
+// BindingUpdate is a distilled representation of the typical kinds of changes a destination system
+// will care about in response to a new binding or change to an existing binding.
+type BindingUpdate struct {
+	// NewProjections is the projections of a proposed materialization spec that are included in the
+	// field selection but there are no fields in the materialized resource for.
+	NewProjections []pf.Projection
+
+	// NewlyNullableFields is the fields in the endpoint that should now be made nullable, that
+	// weren't before. They may need to be nullable because a materialized field is now nullable and
+	// wasn't before, or because they exist as non-nullable in the destination and aren't included
+	// in the materialization's field selection.
+	NewlyNullableFields []EndpointField
```

Comment:
I did it this way so we can remove nullability constraints for fields that we don't have projections for, for example if a column exists in a table that is non-nullable but we aren't materializing a field for it when the `Apply` comes through. In that case we don't have a projection available. I've seen this happen a couple of times when users have re-materialized to tables that were previously created by materializations that had an optional field selected that was always present in the schema, but on the re-materialization they didn't include that optional field.

It certainly does make cases where we have to re-state the column definition more tricky though - although another benefit is that a user could have a pre-created column with certain characteristics (like a different encoding than we'd use for a new projection) and there'd be no risk of us blowing that away if a `NOT NULL` is dropped.

Reply:
Adding to that...I thought about having both the `EndpointField` _and_ a `pf.Projection` if it was available, but then decided that since it has to work with just the `EndpointField` anyway we might as well go with that all the time.